### PR TITLE
feat: add custom field for class name @W-13013140@

### PR DESCRIPTION
**Pull request by joeluong-sfcc**
This PR exposes a custom field, `x-salesforce-sdk-class-name` within the RAML file that is used for the class name for SDK generation. Previously, the class name in the SDK was defined by using the title field. Developers and Technical writers would update the title in the RAML files as they are also used for client facing documentation, leading to unintended downstream impacts onto the SDKs. To help combat these breaking changes, `x-salesforce-sdk-class-name` can be defined explicitly in the RAML file to be used for SDK class name instead of title. If the custom field does not exist, the SDK will use the title by default.

This also allows Isomorphic SDK hot fix to be reverted once the SDK is updated to consume the new RAML toolkit and the Shopper Context RAML is updated to use the custom field: https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic/pull/120

The main logic for this change is located in `src/generate/apiModel.ts`.

### CODE REVIEW

File: src/generate/apiModel.test.ts [hunk 1]
```diff
@@ -102,6 +112,46 @@ describe("ApiModel tests", () => {
       "Cannot update the name before the model is loaded"
     );
   });
+
+  it("uses title for class name if custom field is not set", async () => {
+    const api = new ApiModel("VALID", path.dirname(validRamlFile));
+    await api.init();
+    expect(api.name).to.deep.equal(new Name("Shop API"));
```
Comment (tarcang) [inline src/generate/apiModel.test.ts:119]: Why did you add 'deep' here, is it necessary?

**Review by p-biyani — APPROVED** [review]

File: src/generate/apiModel.ts [hunk 2]
```diff
@@ -72,9 +72,23 @@ export class ApiModel extends ApiMetadata {
...
+
+    // If user defines custom class name under `x-salesforce-sdk-class-name`, use that instead
+    const customClassNameArr = this.model.encodes.customDomainProperties.filter(
+      (customProperty) => customProperty.name.toString() === CUSTOM_NAME_FIELD
     );
+    if (customClassNameArr.length > 0) {
+      // NOTE: there can be multiple instances of `x-salesforce-sdk-class-name` defined, we take the value of the first one
+      const customClassName = customClassNameArr[0].extension
+        .toString()
+        .split("=")[1];
```
Comment (tarcang) [inline src/generate/apiModel.ts:87]: What is the purpose of split('=') here? How is the '=' symbol used while getting the custom title?

**Review by unandyala — APPROVED** [review]
we should post in cc-mercury-dev-internal channel so that they can start using this field

File: src/generate/apiModel.ts [hunk 3]
```diff
@@ -72,9 +72,23 @@ export class ApiModel extends ApiMetadata {
...
       throw new Error("Cannot update the name before the model is loaded");
     }
-    this.name = new Name(
-      (this.model.encodes as model.domain.WebApi)?.name.value()
+
+    // class name is defaulted to title
+    let className = (this.model.encodes as model.domain.WebApi)?.name.value();
+
+    // If user defines custom class name under `x-salesforce-sdk-class-name`, use that instead
+    const customClassNameArr = this.model.encodes.customDomainProperties.filter(
```
Comment (unandyala) [inline src/generate/apiModel.ts:80]: [NIT] extract the name resolution into its own function/method

File: testResources/raml/custom-field/custom-field.raml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#%RAML 1.0
+
+title: Shop API
+x-salesforce-sdk-class-name: Custom Shop API Name
```
Comment (bredmond-sf) [inline testResources/raml/custom-field/custom-field.raml:4]: In RAML this would be `(salesforce-sdk-class-name)`. https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md/#annotations
as opposed to https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md/#x-other

and what if you followed our generator convention of:
```
(commerce-sdk):
  class-name: Custom Shop API Name
```

**Review by bredmond-sf — CHANGES_REQUESTED** [review]

File: src/generate/apiModel.ts (hunk 2 above)
Comment (joeluong-sfcc) [reply src/generate/apiModel.ts]: There was a better way to get the custom field so I updated this part, but originally the string we were splitting on looked like: `scalar_1:http://www.w3.org/2001/XMLSchema#string=Custom Shop API title`, and we'd split on the equals sign to grab the `Custom Shop API title`

File: testResources/raml/custom-field/custom-field.raml (hunk 4 above)
Comment (joeluong-sfcc) [reply testResources/raml/custom-field/custom-field.raml]: Done! You can see this in the example test file `testResources/raml/custom-field/custom-field.raml`

File: src/generate/apiModel.test.ts (hunk 1 above)
Comment (joeluong-sfcc) [reply src/generate/apiModel.test.ts:119]: I believe that `deep` recursively goes through the object to ensure that all properties are equal and structure of the object is the same. I just tested `expect(api.name).to.equal(new Name("Shop API"));` without deep and the unit test fails

**Review by unandyala — APPROVED** [review]
